python/trifinger_simulation/gym_wrapper/envs/cube_trajectory_env.py

```diff
@@ -0,0 +1,341 @@
+                    ob['achieved_goal'],
+                    ob['desired_goal'],
+                    info,
+                )
+        """
+        # This is just some sanity check to verify that the given desired_goal
+        # actually matches with the active goal in the trajectory.
+        active_goal = np.asarray(
+            mct.get_active_goal(
+                self.info["trajectory"], self.info["time_index"]
+            )
+        )
+        assert np.all(active_goal == desired_goal)
+
+        return -mct.evaluate_state(
+            info["trajectory"], info["time_index"], achieved_goal
+        )
+
+    def step(self, action):
+        """Run one timestep of the environment's dynamics.
+
+        When end of episode is reached, you are responsible for calling
+        ``reset()`` to reset this environment's state.
+
+        Args:
+            action: An action provided by the agent (depends on the selected
+                :class:`ActionType`).
+
+        Returns:
+            tuple:
+
+            - observation (dict): agent's observation of the current
+              environment.
+            - reward (float): amount of reward returned after previous action.
+            - done (bool): whether the episode has ended, in which case further
+              step() calls will return undefined results.
+            - info (dict): info dictionary containing the current time index.
+        """
+        if self.platform is None:
+            raise RuntimeError("Call `reset()` before starting to step.")
+
+        if not self.action_space.contains(action):
+            raise ValueError(
+                "Given action is not contained in the action space."
+            )
+
+        num_steps = self.step_size
+
+        # ensure episode length is not exceeded due to step_size
+        step_count_after = self.step_count + num_steps
+        if step_count_after > mct.EPISODE_LENGTH:
+            excess = step_count_after - mct.EPISODE_LENGTH
+            num_steps = max(1, num_steps - excess)
+
+        reward = 0.0
+        for _ in range(num_steps):
+            self.step_count += 1
+            if self.step_count > mct.EPISODE_LENGTH:
+                raise RuntimeError("Exceeded number of steps for one episode.")
+
+            # send action to robot
+            robot_action = self._gym_action_to_robot_action(action)
+            t = self.platform.append_desired_action(robot_action)
+            self.info["time_index"] = t
+
+            # update goal visualization
+            if self.visualization:
+                goal_position = mct.get_active_goal(self.info["trajectory"], t)
+                self.goal_marker.set_state(goal_position, (0, 0, 0, 1))
+
+            # Use observations of step t + 1 to follow what would be expected
+            # in a typical gym environment.  Note that on the real robot, this
+            # will not be possible
+            observation = self._create_observation(t + 1)
```

Comment:
t+1 should be the right choice here as you would typically send action a_(t) and want to get back o_(t+1) and r_(t+1)